File: java/hackerrank/tree-level-order-traversal.java
import java.util.*;
import java.io.*;

class Node {
    Node left;
    Node right;
    int data;
    
    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}

class Solution {
	/* 
    
    class Node 
    	int data;
    	Node left;
    	Node right;
	*/
	public static void levelOrder(Node root) {
        List<List<Integer>> result = new ArrayList();
        int depth = 0;
        result = findLevels(root, result, depth);
        for (int i = 0; i < result.size(); i++) {
            List<Integer> l = result.get(i);
            for (int j = 0; j < l.size(); j++) {
                int v = l.get(j);
                System.out.print(v);
                if (j < l.size() - 1) {
                    System.out.printf(" ");
                }
            }
            if (i < result.size() - 1) {
                System.out.print(" ");
            }
        }
    }

    public static List<List<Integer>> findLevels(Node root, List<List<Integer>> result, int depth) {
        if (root == null) {
            return result;
        }
        if (result.size() <= depth) {
            List<Integer> newRow = new ArrayList();
            result.add(newRow);
        }
        result.get(depth).add(root.data);
        if (root.left != null) {
            result = findLevels(root.left, result, depth + 1);
        }
        if (root.right != null) {
            result = findLevels(root.right, result, depth + 1);
        }
        return result;
    }
	public static Node insert(Node root, int data) {
        if(root == null) {
            return new Node(data);
        } else {
            Node cur;
            if(data <= root.data) {
                cur = insert(root.left, data);
                root.left = cur;
            } else {
                cur = insert(root.right, data);
                root.right = cur;
            }
            return root;
        }
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int t = scan.nextInt();
        Node root = null;
        while(t-- > 0) {
            int data = scan.nextInt();
            root = insert(root, data);
        }
        scan.close();
        levelOrder(root);
    }	
}

File: python/hackerrank/tree-level-order-traversal.py
class Node:
    def __init__(self, info): 
        self.info = info  
        self.left = None  
        self.right = None 
        self.level = None 

    def __str__(self):
        return str(self.info) 

class BinarySearchTree:
    def __init__(self): 
        self.root = None

    def create(self, val):  
        if self.root == None:
            self.root = Node(val)
        else:
            current = self.root
         
            while True:
                if val < current.info:
                    if current.left:
                        current = current.left
                    else:
                        current.left = Node(val)
                        break
                elif val > current.info:
                    if current.right:
                        current = current.right
                    else:
                        current.right = Node(val)
                        break
                else:
                    break
import sys
"""
Node is defined as
self.left (the left child of the node)
self.right (the right child of the node)
self.info (the value of the node)
"""
def levelOrder(root):
    #Write your code here
        result = []
        depth = 0
        levels = findLevels(root, result, depth)
        for i in range(len(levels)):
            l = levels[i]
            for j in range(len(l)):
                v = l[j]
                sys.stdout.write(str(v))
                if j < len(l) - 1:
                    sys.stdout.write(' ')
            if i < len(levels) - 1:
                sys.stdout.write(' ')
                    
def findLevels(root, result, depth):
    if root is None:
        return result
    if len(result) <= depth:
        result.append([])
    result[depth].append(root.info)
    if root.left is not None:
        result = findLevels(root.left, result, depth + 1)
    if root.right is not None:
        result = findLevels(root.right, result, depth + 1)
    return result


tree = BinarySearchTree()
t = int(input())

arr = list(map(int, input().split()))

for i in range(t):
    tree.create(arr[i])

levelOrder(tree.root)
